File: Verify Preorder Serialization of a Binary Tree.py
class Item:
    def __init__(self, vertex):
        self.__vertex = vertex
        self.__status = 0
    def addStatus(self):
        self.__status += 1
    @property
    def finished(self):
        return self.__status == 2
    @property
    def vertex(self):
        return self.__vertex

class Solution(object):
    def isValidSerialization(self, preorder):
        if not preorder:return False
        array = preorder.split(',')
        if len(array) == 1 and array[0] == '#':return True
        stack = []
        node_array = [Item(i) for i in array]
        stack.append(node_array[0])
        for node in node_array[1:]:
            if not stack:return False
            stack[-1].addStatus()
            if stack[-1].finished:stack.pop()
            if node.vertex != '#':stack.append(node)
        return stack == []

File: Verify Preorder Serialization of a Binary Tree.cpp
class Item {
public:
    Item(string _vertex): vertex(_vertex), status(0) {}
    bool finished() {
        return status == 2;
    }
    string &getVertex() {
        return vertex;
    }
    void addStatus() {
        ++status;
    }
private:
    string vertex;
    int status;
};

void stringSplit(string input, char split, vector<Item*>& afterSplit) {
    int start = 0, end = 0;
    for (int i = 0;i < input.size(); ++i) {
        if (input[i] == split) {
            end = i;
            string substring = input.substr(start, end - start);
            Item *item = new Item(substring);
            afterSplit.push_back(item);
            start = i + 1;
        }
    }
    if (start < input.size()) {
        string substring = input.substr(start, input.size() - start);
        Item *item = new Item(substring);
        afterSplit.push_back(item);
    }
}

class Solution {
public:
    bool isValidSerialization(string preorder) {
        vector<Item*> input;
        stringSplit(preorder, ',', input);
        if (input.size() == 0) {
            return false;
        }
        if (input.size() == 1 && input[0]->getVertex() == "#") {
            return true;
        }
        stack<Item*> istack;
        vector<Item*>::iterator it = input.begin();
        istack.push(*it);
        ++it;
        for (;it != input.end(); ++it) {
            if (istack.empty()) {
                return false;
            }
            Item* top = istack.top();
            top->addStatus();
            if (top->finished()) {
                istack.pop();
                delete top;
            }
            if ((*it)->getVertex() != "#") {
                istack.push(*it);
            }
        }
        return istack.empty();
    }
};